test/src/unit-capi-smoke-test.cc

```diff
@@ -0,0 +1,757 @@
+      } else {
+        write_buffer =
+            (uint64_t*)tdb_malloc(range * (range - 1) * sizeof(uint64_t*));
+      }
+      d_write_buffers.emplace_back(write_buffer);
+    }
+
+    // Fill the write buffers with values.
+    auto dims_iter = test_dims.begin();
+    for (uint64_t* const d_write_buffer : d_write_buffers) {
+      for (uint64_t i = 0; i < dimension_ranges[i]; i++) {
+        d_write_buffer[i] = i;
+      }
+      write_buffer_size = dim_buffer_len * sizeof(uint64_t);
+
+      write_query_buffers.emplace_back(
+          dims_iter->name_,
+          d_write_buffer,
+          &(write_buffer_size),
+          nullptr,
+          nullptr);
+      dims_iter = std::next(dims_iter, 1);
+    }
+
+    // Execute the write query.
+    write(array_name, write_query_buffers, write_order, encryption_type);
+
+    // Free the dimension write buffers vector
+    d_write_buffers.clear();
+    d_write_buffers.shrink_to_fit();
+  } else {
+    // Execute the write query.
+    write(array_name, write_query_buffers, write_order, encryption_type);
+  }
+
+  // Define the read query buffers for "a".
+  vector<test_query_buffer_t> read_query_buffers;
+
+  uint64_t a_read_buffer_size;
+  int32_t* a_read_buffer = nullptr;
+
+  if (test_attr.cell_val_num_ == TILEDB_VAR_NUM) {
+    a_read_buffer_size =
+        total_cells * 2 * tiledb_datatype_size(test_attr.type_);
+    a_read_buffer = (int32_t*)tdb_malloc(a_read_buffer_size);
+    for (uint64_t i = 0; i < total_cells * 2; i++) {
+      a_read_buffer[i] = 0;
+    }
+  } else {
+    a_read_buffer_size = total_cells * tiledb_datatype_size(test_attr.type_);
+    a_read_buffer = (int32_t*)tdb_malloc(a_read_buffer_size);
+    for (uint64_t i = 0; i < total_cells; i++) {
+      a_read_buffer[i] = 0;
+    }
+  }
+
+  uint64_t a_read_buffer_offset_size = total_cells * sizeof(uint64_t);
+  uint64_t* a_read_buffer_offset =
+      (uint64_t*)tdb_malloc(a_read_buffer_offset_size);
+  for (uint64_t i = 0; i < total_cells; i++) {
+    a_read_buffer_offset[i] = 0;
+  }
+
+  // Variable buffer is set only when the attribute is var-sized.
+  if (test_attr.cell_val_num_ == TILEDB_VAR_NUM) {
+    read_query_buffers.emplace_back(
+        test_attr.name_,
+        a_read_buffer_offset,
+        &a_read_buffer_offset_size,
+        a_read_buffer,
+        &a_read_buffer_size);
+  } else {
+    read_query_buffers.emplace_back(
+        test_attr.name_, a_read_buffer, &a_read_buffer_size, nullptr, nullptr);
+  }
+
+  uint64_t subarray_len = 2 * test_dims.size();
+  uint64_t subarray_size = subarray_len * sizeof(uint64_t);
+  uint64_t* subarray_full = (uint64_t*)tdb_malloc(subarray_size);
+  for (uint64_t i = 0; i < subarray_len; i += 2) {
+    const uint64_t min_range = ((uint64_t*)(test_dims[i / 2].domain_))[0];
+    const uint64_t max_range = ((uint64_t*)(test_dims[i / 2].domain_))[1];
+    subarray_full[i] = min_range;
+    subarray_full[i + 1] = max_range;
```

Comment:
this is pedantic, but could you refactor the for-loop to use `++i` rather than `i += 2` and then readjust the logic accordingly?

e.g.
```
for (uint64_t i = 0; i < subarray_len; ++i) {
    const uint64_t min_range = ((uint64_t*)(test_dims[i].domain_))[0];
    const uint64_t max_range = ((uint64_t*)(test_dims[i].domain_))[1];
    subarray_full[(i * 2)] = min_range;
    subarray_full[(i * 2) + 1] = max_range;
}
```

the motivation is that cpu division operations are typically twice as expensive as multiplications. it doesnt really matter in a test, but we follow this philosophy in the core so let's stay consistent here

Reply:
Alternatively consider adding second idx as 'it(imes)2', while keeping i, adjust them both, ++i, it2+=2, 
thus adding one add per loop iteration but avoiding both of either the divisions or the multiplications... ?

Reply:
This logic makes total sense to me. More sense in fact than my original code. However, I'm currently getting a segfault when trying to implement it. Looking into that now...not sure why this change would make that happen...